### PR TITLE
ENH: add c-imported modules for freeze analysis in np.random

### PR DESCRIPTION
**Pull request by KmolYuan**
New module `numpy.random` in `1.17.0` caused `ModuleNotFoundError` for module-freeze analysis project like PyInstaller.

```bash
  File "c:\python37-x64\lib\site-packages\PyInstaller\loader\pyimod03_importers.py", line 627, in exec_module
    exec(bytecode, module.__dict__)
  File "site-packages\numpy\__init__.py", line 150, in <module>
  File "c:\python37-x64\lib\site-packages\PyInstaller\loader\pyimod03_importers.py", line 627, in exec_module
    exec(bytecode, module.__dict__)
  File "site-packages\numpy\random\__init__.py", line 180, in <module>
  File "mtrand.pyx", line 1, in init numpy.random.mtrand
ModuleNotFoundError: No module named 'numpy.random.common'
```

So I added some "hooks" for that.

### CODE REVIEW

**Review by seberg — APPROVED** [review]
Looks good to me, thanks. I would have had a tendency to put these last, but it odes not really matter.

**Comment by charris** [comment]
What is module-freeze analysis? I'm a bit chary of adding what extraneous imports, is this a NumPy problem or a problem with the analysis code?

**Comment by eric-wieser** [comment]
I'm with @charris here - this looks ok for now to unblock freezing, but it's at odds with the type of change that @hmaarrfk has been trying to make to speed up import times. Would putting these imports in `if False` be enough for the freezer to find them?

**Comment by seberg** [comment]
The `if False` is an interesting idea. Yeah, had for a bit wondered about the `__getattr__`, etc. as well. We do have one other identical pattern in the code right now.

This must be a general problem with cython files, etc. I wonder if it is not a bit of an issue with the freezing tools being too agressive.

**Comment by seberg** [comment]
Checking the pyinstaller docs, it sounds like running python with `--debug=imports` while building might already give you all necessary imports reliably, could be the right approach to advocate that?

**Comment by eric-wieser** [comment]
This sounds like the issue is that pyinstaller is unable to find imports within cython files, which is arguably an upstream problem.

**Comment by seberg** [comment]
I agree it is an upstream problem (either with cython, or probabl pyinstaller). But if it affects enough use cases, I am willing to compromise, since it does not seem too harmfll as well.

For reference, the other place we did this before is gh-12871 (although those are not cython modules, so the failure mode is maybe a tiny bit different, effectively both cython and C Extension modules are similar in this regard).

**Comment by hmaarrfk** [comment]
Two questions:
1. How will we make sure this kind of thing doesn't happen again? Is there a test for pyfreezers?
2. Is it possible to add these to the submodule `__init__` files and not the top level one?

**Comment by hmaarrfk** [comment]
I don't think this conflicts with the desire for faster import times per-say. Anywho, I think the location of these imports shouldn't be the top level `__init__.py`

Rather, they should be placed in `numpy/random/__init__.py` with the appropriate comment

```
diff --git a/numpy/random/__init__.py b/numpy/random/__init__.py
index 2d495d67e..4b242dfa6 100644
--- a/numpy/random/__init__.py
+++ b/numpy/random/__init__.py
@@ -207,6 +207,11 @@ def __RandomState_ctor():
     return RandomState(seed=0)


+# This isn't actually used in __all__, but
+# PyInstaller requires it to find c-imported modules
+from . import common
+from . import bounded_integers
+from . import entropy
 from numpy._pytesttester import PytestTester
 test = PytestTester(__name__)
 del PytestTester
```

**Comment by eric-wieser** [comment]
@hmaarrfk, I think you're misreading the filename, the change you're requesting is what this PR already does

**Comment by seberg** [comment]
xref gh-14163

**Comment by seberg** [comment]
I am somewhat guessing we are likely coming to "meehh, fine if there is no good workaround in PyInstaller"?

@KmolYuan can you try the `if False:` block? I am not sure how PyInstaller works, would be good to know if it does the trick, right now I am not sure what approach PyInstaller uses to discover imports.

**Comment by eric-wieser** [comment]
I'd like to block this on an upstream bug being filled, so we at least have something to reference in a comment

**Comment by KmolYuan** [comment]
Some ways to add import libraries in the freeze analysis of PyInstaller:

1. `import` statements in the Python scripts, which can find the name of the libraries (also use the `ldd` command). **PyInstaller will not check if the code can be reached, unless add them into exclude module list manually.**
1. Official settings of PyInstaller, needs to add a PR to their repository every times.
1. User options from configuration file or command line arguments.

PyInstaller doesn't check Cython sources, so import a Python module (`import` statements) in Cython will not detect.

**Comment by rkern** [comment]
I'd rather see a PR against PyInstaller to add the appropriate hooks. There are already `numpy`-specific hooks to pick up some of the external DLLs provided by the PyPI binary packages. Other packages with Cython imports appear to use this mechanism.

https://github.com/pyinstaller/pyinstaller/tree/develop/PyInstaller/hooks

**Comment by eric-wieser** [comment]
https://github.com/pyinstaller/pyinstaller/issues/1674 tracks finding cython extensions

**Comment by charris** [comment]
The consensus seems to be that this should be fixed upstream. Adding a hook seems the quick way to go, there are already two numpy related hooks, do we need to add a `hook-numpy.random.py` or can we modify one of the existing ones? I don't know how that is supposed to work. Who should do it?

**Comment by charris** [comment]
We should also deal with gh-12871 if we add an upstream hook.

**Comment by rkern** [comment]
One would write a `hook-numpy.random.py`, yes. Not sure how different versions with different contents are dealt with.

**Comment by charris** [comment]
I'm putting this in as a workaround until there is a better fix. The downside is it removes the pressure on the affected packages to make a fix. The upside is it removes the pressure on us.

**Comment by charris** [comment]
Thanks @KmolYuan .

**Pull request merged**
